### ISSUE TITLE
`unsafe_derive_deserialize` triggers on "transparent" types

### ISSUE THREAD

**Issue by SLUCHABLUB** [rust-lang/rust-clippy#16319]
### Summary

The raison d'être of this lint is "Deriving `serde::Deserialize` will create a constructor that may violate invariants held by another constructor." however, if a type is an `enum`, or a `struct` with all `pub`lic fields, then there is already a constructor (or multiple) so there can't be any invariants.

```rust
// This struct may have an invariant.
// For example: `bar` is non-negative.
struct Foo {
    bar: i32
}

// However, this struct cannot have any invariants
// since anyone can write: Foo { bar: /* any value */ }
struct Foo {
    pub bar: i32
}
```

### Lint Name

unsafe_derive_deserialize

### Reproducer

I tried this code:

```rust
#![deny(clippy::unsafe_derive_deserialize)]
#![allow(dead_code)]

extern crate serde; // 1.0.228

#[derive(serde::Deserialize)]
struct Foo {
    pub bar: i32,
}

impl Foo {
    unsafe fn oof() {}
}
```

I saw this happen:

```
error: you are deriving `serde::Deserialize` on a type that has methods using `unsafe`
 --> src/lib.rs:6:10
  |
6 | #[derive(serde::Deserialize)]
  |          ^^^^^^^^^^^^^^^^^^
  |
  = help: consider implementing `serde::Deserialize` manually. See https://serde.rs/impl-deserialize.html
  = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.92.0/index.html#unsafe_derive_deserialize
note: the lint level is defined here
 --> src/lib.rs:1:9
  |
1 | #![deny(clippy::unsafe_derive_deserialize)]
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  = note: this error originates in the derive macro `serde::Deserialize` (in Nightly builds, run with -Z macro-backtrace for more info)
```

I expected to see this happen:

nothing

### Version

```text
rustc 1.92.0 (ded5c06cf 2025-12-08)
binary: rustc
commit-hash: ded5c06cf21d2b93bffd5d884aa6e96934ee4234
commit-date: 2025-12-08
host: x86_64-unknown-linux-gnu
release: 1.92.0
LLVM version: 21.1.3
```

**Comment by ada4a** [comment]
Isn't having an unsafe method on a struct/enum with all-public fields a problem on its own? Maybe we should have a lint against that?

**Comment by SLUCHABLUB** [comment]
Not necessarily. In my case the type was an `enum` with an unsafe constructor that constructs it from a discriminant.

Here is another minimal example:
```rust
#![deny(clippy::unsafe_derive_deserialize)]
#![allow(dead_code)]

extern crate serde; // 1.0.228

#[derive(serde::Deserialize)]
enum SmallInt {
    Zero,
    One,
    Two,
}

impl SmallInt {
    unsafe fn from_u8_unchecked(_: u8) -> Self {
        todo!()
    }
}
```

**Comment by ada4a** [comment]
Hmm, there are 3 points here I think:
- in the case of field-less enums (and structs, I guess), we shouldn't lint at all
- for items with all-public fields, we should lint, but not with `unsafe_derive_deserialize` (but rather the "why do you have all-public fields and unsafe methods" lint
- we should probably ignore unsafe _constructors_ for the purposes of both of those lints, as, well, deserialization is really just an unsafe constructor as well

WDYT?
